### PR TITLE
NO-ISSUE: e2e testing in VMs test harness

### PR DESCRIPTION
**Pull request by mangelajo**
Adds the foundations for `test/e2e` and a framework to run, and interact with VMs (console and ssh). The VMs are pre-configured to register to the local-kind deployment.

### CODE REVIEW

File: test/e2e/agent_test.go [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+	var (
+		testVM vm.BootcVM
+	)
+
+	BeforeEach(func() {
+
+		currentWorkDirectory, err := os.Getwd()
+
+		Expect(err).ToNot(HaveOccurred())
+		params := vm.NewVMParameters{
```
Comment (mangelajo) [inline test/e2e/agent_test.go:31]: Ok, next step, I should simplify all this parameter passing and solidify
a proper e2e test harness.

File: hack/build_rpms.sh [hunk 2]
```diff
@@ -4,11 +4,10 @@ if [ -f /etc/debian_version ]; then
     echo "Building RPMs on a non-rpm based system via podman..."
     cat >bin/build_rpms.sh <<EOF
 #!/usr/bin/env bash
-dnf install -y make packit golang openssl-devel
 cd /work
 ./hack/build_rpms_packit.sh
 EOF
-    podman run --rm -t -v $(pwd):/work fedora:39 bash /work/bin/build_rpms.sh
+    podman run --rm -t -v $(pwd):/work quay.io/flightctl/ci-rpm-builder:latest bash /work/bin/build_rpms.sh
```
Comment (mangelajo) [inline hack/build_rpms.sh:10]: This saves 2+ minutes of build time, I need to upload the containerfile for this and setup CI.

File: .github/workflows/lint.yaml [hunk 3]
```diff
@@ -27,24 +27,12 @@ jobs:
...
+        uses: ./.github/actions/setup-dependencies
 
-      - name: Check that generated files have been updated
+      - name: Check that generated files have been updated and that go.mod is tidy
         run: |
           make generate
-          git diff --exit-code
-
-      - name: Check that go.mod is tidy
-        run: |
```
Comment (mangelajo) [inline .github/workflows/lint.yaml:47]: To handle this issue: https://github.com/flightctl/flightctl/actions/runs/8632458059/job/23663198421?pr=205#step:6:158

Seems like make generate massages the go.mod/sum for some reason in CI

**Review by sagidayan — COMMENTED** [review]
Had a small nit. 

Looks great so far. I'm still going over it...
Also, ill try to run it locally.

We definitely need more eyes on this.

Good job mate!

File: test/harness/e2e/vm/cloudinit.go [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+
+import (
+	"errors"
+	"fmt"
+	"os"
+	"os/exec"
+	"path/filepath"
+)
+
+func (b *TestVM) ParseCloudInit() (err error) {
```
[the remark this answers is not in the document]
Comment (mangelajo) [reply test/harness/e2e/vm/cloudinit.go:11]: True, let's clean that up

Thanks :)

File: test/harness/e2e/vm/README.md [hunk 5]
```diff
@@ -0,0 +1,2 @@
+This package is inspired from the code in https://gitlab.com/bootc-org/podman-bootc-cli/-/tree/main/pkg/vm
+as a very quick way to gent a bootc VM running for testing purposes.
```
Comment (oglok) [inline test/harness/e2e/vm/README.md:2]: s/gent/generate?

File: test/e2e/agent_test.go [hunk 6]
```diff
@@ -0,0 +1,69 @@
...
+	RunSpecs(t, "Agent E2E Suite")
+}
+
+var _ = Describe("VM Agent behavior", func() {
+	var (
+		testVM vm.TestVMInterface
+	)
+
+	BeforeEach(func() {
+
```
Comment (oglok) [inline test/e2e/agent_test.go:30]: nit: remove this blank line

File: test/e2e/agent_test.go [hunk 7]
```diff
@@ -0,0 +1,69 @@
...
+		currentWorkDirectory, err := os.Getwd()
+
```
Comment (oglok) [inline test/e2e/agent_test.go:32]: nit: remove this blank line

File: test/harness/e2e/vm/cloudinit.go [hunk 8]
```diff
@@ -0,0 +1,43 @@
...
+
+import (
+	"errors"
+	"fmt"
+	"os"
+	"os/exec"
+	"path/filepath"
+)
+
+func (b *TestVM) ParseCloudInit() error {
```
Comment (oglok) [inline test/harness/e2e/vm/cloudinit.go:11]: why "b" here? the rest of the receiver params are "v"s in other files. Just curious if it was a typo.

File: test/harness/e2e/vm/vm.go [hunk 9]
```diff
@@ -0,0 +1,133 @@
...
+	WaitForSSHToBeReady() error
+	SSHCommand(inputArgs []string) *exec.Cmd
+	RunSSH(inputArgs []string, stdin *bytes.Buffer) (*bytes.Buffer, error)
+	Exists() (bool, error)
+	ReadConsole() string
+	GetConsoleOutput() string
+}
+
+func (v *TestVM) WaitForSSHToBeReady() error {
+	timeout := 60 * time.Second
```
Comment (oglok) [inline test/harness/e2e/vm/vm.go:47]: const?

File: test/harness/e2e/vm/vm.go [hunk 10]
```diff
@@ -0,0 +1,133 @@
...
+	elapsed := 0 * time.Second
```
Comment (oglok) [inline test/harness/e2e/vm/vm.go:48]: const?

**Review by oglok — COMMENTED** [review]
It looks good, just minor nits.

Some of the code seems to be a wrapper on libvirt functions, like shutdown, destroy, etc. I wonder if that's only to make it functions of the VM struct or could be enough just to call the libvirt library and getting domain_name (i.e.) as input args.

File: test/harness/e2e/vm/vm.go (hunk 9 above)
Comment (mangelajo) [reply test/harness/e2e/vm/vm.go]: Done.

**Comment by mangelajo** [comment]
> It looks good, just minor nits.
> 
> Some of the code seems to be a wrapper on libvirt functions, like shutdown, destroy, etc. I wonder if that's only to make it functions of the VM struct or could be enough just to call the libvirt library and getting domain_name (i.e.) as input args.

The reason for the abstraction is that in the original bootc version they also support MacOS, which we could eventually try to support too, it's probably not super difficult. They call bare qemu with flags.

**Comment by mangelajo** [comment]
> It looks good, just minor nits.
> 
> Some of the code seems to be a wrapper on libvirt functions, like shutdown, destroy, etc. I wonder if that's only to make it functions of the VM struct or could be enough just to call the libvirt library and getting domain_name (i.e.) as input args.

Also to make it easier to use from testing.

**Review by oglok — APPROVED** [review]

**Pull request merged**